### PR TITLE
amp sigma noise

### PR DESCRIPTION
**Pull request by a-quelle**
Added support for this parameter from the `NoiseModel` for emu-mps

### CODE REVIEW

File: emu_base/pulser_adapter.py [hunk 1]
```diff
@@ -169,10 +183,8 @@ def _extract_omega_delta_phi(
                     locals_a_d_p[q_id]["phase"][t1] + locals_a_d_p[q_id]["phase"][t2]
                 ) / 2.0
             # omegas at different times need to have the laser waist applied independently
-            if t1 in global_times:
-                omega_1 *= waist_factors
-            if t2 in global_times:
-                omega_2 *= waist_factors
+            omega_1 *= global_times_to_amp_factors.get(t1, 1.0)
```
Comment (pablolh) [inline emu_base/pulser_adapter.py:186]: Nothing related to this change, but could you move those two lines outside of the `q_pos, q_id` loop, it's weird now cuz they are defined in the loop and used outside
```
t1 = math.floor(t)
t2 = math.ceil(t)
```

File: emu_base/pulser_adapter.py [hunk 2]
```diff
@@ -131,20 +132,33 @@ def _extract_omega_delta_phi(
...
+        prop_dir = torch.tensor(
+            ch_obj.propagation_dir or [0.0, 1.0, 0.0], dtype=torch.float64
+        )
+        prop_dir /= prop_dir.norm()
+        sigma_factor = 1.0 + amp_sigma**2 * torch.max(torch.tensor(0), torch.randn(1))
+        for slot in ch_samples.slots:
+            factors = (
+                torch.tensor(
+                    [
+                        math.exp(-torch.norm(perp_dist(x, prop_dir) / laser_waist) ** 2)
```
[the remark this answers is not in the document]
Comment (sgrava) [reply emu_base/pulser_adapter.py:152]: I agree, it should be just a scalar, right? The distance from the axis of the propagation direction

[the remark this answers is not in the document]
Comment (a-quelle) [reply emu_base/pulser_adapter.py:160]: yes, this is the perpendicular component of the position, the distance would be the norm of this vector. I moved the norm from line 152 into `perp_dist`

File: emu_base/pulser_adapter.py [hunk 3]
```diff
@@ -131,20 +132,33 @@ def _extract_omega_delta_phi(
...
+                        for x in qubit_positions
+                    ]
+                )
+                if laser_waist and ch_obj.addressing == "Global"
+                else torch.ones(len(sequence.register.qubit_ids))
+            )
+            factors *= sigma_factor
+            for i in range(slot.ti, slot.tf):
+                global_times_to_amp_factors[i] = factors
```
[the remark this answers is not in the document]
Comment (a-quelle) [reply emu_base/pulser_adapter.py]: indeed, I changed the name

File: emu_base/pulser_adapter.py [hunk 4]
```diff
@@ -131,20 +132,33 @@ def _extract_omega_delta_phi(
...
     for ch, ch_samples in samples.channel_samples.items():
-        if samples._ch_objs[ch].addressing == "Global":
-            for slot in ch_samples.slots:
-                global_times |= set(i for i in range(slot.ti, slot.tf))
+        ch_obj = samples._ch_objs[ch]
+        prop_dir = torch.tensor(
+            ch_obj.propagation_dir or [0.0, 1.0, 0.0], dtype=torch.float64
+        )
+        prop_dir /= prop_dir.norm()
+        sigma_factor = 1.0 + amp_sigma**2 * torch.max(torch.tensor(0), torch.randn(1))
```
[the remark this answers is not in the document]
Comment (a-quelle) [reply emu_base/pulser_adapter.py]: indeed, good call, I will fix this and also update the tests

File: emu_base/pulser_adapter.py (hunk 1 above)
Comment (a-quelle) [reply emu_base/pulser_adapter.py]: done

File: emu_base/pulser_adapter.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (a-quelle) [reply emu_base/pulser_adapter.py]: I added some comments, let me know if you think it's enough

File: emu_base/pulser_adapter.py [hunk 5]
```diff
@@ -117,34 +120,63 @@ def _extract_omega_delta_phi(
...
+        ch_obj = samples._ch_objs[ch]
+        prop_dir = torch.tensor(
+            ch_obj.propagation_dir or [0.0, 1.0, 0.0], dtype=torch.float64
+        )
+        prop_dir /= prop_dir.norm()
+
+        # each channel has a noise on its laser amplitude
+        # we assume each channel has the same noise amplitude currently
+        # the hardware currently has only a global channel anyway
+        sigma_factor = torch.max(
```
Comment (pablolh) [inline emu_base/pulser_adapter.py:153]: I find a bit sad that you had to modify bitstring counts on non-noisy end-to-end tests. Maybe to avoid changing the state of the random number generator, I personally wouldn't be shocked by
```
sigma_factor = 1.0 if amp_sigma == 0.0 else max(
            0, 1.0 + amp_sigma**2 * torch.randn(1).item()
        )
```

It's also a sanity test - if the bitstrings do not go back to normal then we have another problem?

**Review by pablolh — COMMENTED** [review]
Latest change looks good to me, one last comment to minimize the scope of the diff regarding bitstrings

File: emu_base/pulser_adapter.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sgrava) [reply emu_base/pulser_adapter.py]: Nice!

File: emu_base/pulser_adapter.py (hunk 5 above)
Comment (a-quelle) [reply emu_base/pulser_adapter.py:153]: good point. I also realized during my sleep that the scale factor should be `amp_sigma` not the square.

File: emu_base/pulser_adapter.py [hunk 6]
```diff
@@ -117,34 +120,63 @@ def _extract_omega_delta_phi(
...
+    qubit_positions = _get_qubit_positions(sequence.register)
 
-    if laser_waist:
-        qubit_positions = _get_qubit_positions(sequence.register)
-        waist_factors = torch.tensor(
-            [math.exp(-((x[:2].norm() / laser_waist) ** 2)) for x in qubit_positions]
-        )
-    else:
-        waist_factors = torch.ones(len(sequence.register.qubit_ids))
+    def perp_dist(pos: torch.Tensor, axis: torch.Tensor) -> Any:
```
Comment (kbidzhiev) [inline emu_base/pulser_adapter.py:139]: why you return `Any` ? it seems that `torch.linalg.vector_norm` returns `Tensor`

Comment (a-quelle) [reply emu_base/pulser_adapter.py:140]: it would seem so at first sight, but it actually returns `Any`... Maybe it's an issue with my mypy environment, but it's like that for me.

**Review by pablolh — APPROVED** [review]
